File: Codility/02. Arrays/CyclicRotation/Solution.swift
// https://app.codility.com/programmers/lessons/2-arrays/cyclic_rotation/
// Painless

public func solution(_ A : inout [Int], _ K : Int) -> [Int] {
    let N = A.count
    if N == 0 {
        return A
    }

    let k = K % N
    if k > 0 {
        for _ in 1...k {
            let temp = A[N - 1]
            for j in (0...N - 2).reversed() {
                A[j + 1] = A[j]
            }
            A[0] = temp
        }
    }

    return A
}

File: Codility/02. Arrays/CyclicRotation/Solution.py
# https://app.codility.com/programmers/lessons/2-arrays/cyclic_rotation/
# Painless

from typing import List


def solution(A: List[int], K: int) -> List[int]:
    """Rotate right by slicing the array at the K-mod-length boundary and swapping the two parts."""
    # Time: O(n)   Space: O(n)
    if not A:
        return A

    k = K % len(A)
    return A[-k:] + A[:-k] if k else A[:]


def test() -> None:
    assert solution([3, 8, 9, 7, 6], 3) == [9, 7, 6, 3, 8]
    assert solution([1, 2, 3, 4], 4) == [1, 2, 3, 4]
    assert solution([], 5) == []
    assert solution([5], 7) == [5]
